runtime/src/main/cpp/MemorySharedRefs.cpp

```diff
@@ -126,15 +138,33 @@ void BackRefFromAssociatedObject::releaseRef() {
 }
 
 ObjHeader* BackRefFromAssociatedObject::ref() const {
-  ensureRefAccessible();
+  return konan::tryOrTerminate([this]() { return refOrThrow(); });
```

Comment:
This is better than before, but still suffers from bloating / duplication. Consider templating of entire call tree: ref, addRef, tryAddRef etc. See https://github.com/JetBrains/kotlin-native/tree/vi/ref-shared-holder-exceptions-templating for example